### PR TITLE
Prevent going back to participant page when categorizing

### PR DESCRIPTION
**Pull request by nkdengineer**
### Explanation of Change

Prevent going back to participant page when categorizing
### Fixed Issues

$ https://github.com/Expensify/App/issues/53274
https://github.com/Expensify/App/issues/53318
PROPOSAL:

### Tests

- https://github.com/Expensify/App/issues/53274

1. Login with a new account
2. Create two new workspaces
3. Create a track expense
4. Categorize it
5. Select a category
6. On the confirmation page, click on the back button
7. Verify that the category page is displayed

- https://github.com/Expensify/App/issues/53318 (only Android/IOS app)

Precondition:

- Device font size should be small.
- Account has no workspace.

1. Create a track expense
2. Categorize it
3. On the upgrade page, click on the upgrade button
4. Verify that the success message is center-aligned
- [x] Verify that no errors appear in the JS console

### Offline tests

Same
### QA Steps

// TODO: These must be filled out, or the issue title must include "[No QA]."
- https://github.com/Expensify/App/issues/53274

1. Login with a new account
2. Create two new workspaces
3. Create a track expense
4. Categorize it
5. Select a category
6. On the confirmation page, click on the back button
7. Verify that the category page is displayed

- https://github.com/Expensify/App/issues/53318 (only Android/IOS app)

Precondition:

- Device font size should be small.
- Account has no workspace.

1. Create a track expense
2. Categorize it
3. On the upgrade page, click on the upgrade button
4. Verify that the success message is center-aligned
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I used [JaimeGPT](https://chatgpt.com/g/g-2dgOQl5VM-english-to-spanish-translator-aka-jaimegpt) to get English > Spanish translation. I then posted it in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.ts or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I added [unit tests](https://github.com/Expensify/App/blob/main/tests/README.md) for any new feature or bug fix in this PR to help automatically prevent regressions in this user flow.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/9171fcd0-e17c-4d36-b5a6-e42b3b61da98

Screenshot 2024-12-03 at 11 33 31

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/ca788ce7-c6d6-4ea2-996d-7075b73a2591

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/3bb345d3-e85a-4ce4-8e18-8d054e48d6cf

Screenshot 2024-12-03 at 11 31 55

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/247cac59-14e1-4306-bd68-2eee2c1132da

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/09b30572-4f0b-4ba8-a5ac-b2ee188fe881

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/fb463fc8-0424-426a-86dd-a4c70666f3d4

</details>

### CODE REVIEW

**Comment by trjExpensify** [comment]
Yes, that's what I would expect. If you go back from the category list, then we close the RHP and NOT show a workspace list.

**Comment by nkdengineer** [comment]
@Ollyws All good now.

**Comment by trjExpensify** [comment]
Right, I agree if you go back from the confirmation page it should take you to the category list. If you go back from the category list, it closes the RHP.

**Comment by nkdengineer** [comment]
@Ollyws When we are from the category step we always `closeAndNavigate` to the confirmation step. Then I think it's good to close the RHP in confirmation in this case because if in the confirmation step we click on `category` item to open the category step and click the back button two times, it also closes the RHP.

https://github.com/Expensify/App/blob/ce282d38777e9e0a267e1b6f44bff9c4aace98ee/src/pages/iou/request/step/IOURequestStepCategory.tsx#L130-L133

cc @trjExpensify 

https://github.com/user-attachments/assets/15a36556-bd04-4131-96e4-8e277e2bf533

**Comment by trjExpensify** [comment]
Yeah, my expectation was: 

1. Click "categorize it" 
2. Upgrade to a workspace to unlock categories
3. Choose a category
4. Land on the confirmation page
5. Go back 
6. Choose a category 
7. Go back 
5. Close the RHP 

Whereas, what appears to be happening is: 

1. Click "categorize it" 
2. Upgrade to a workspace to unlock categories
3. Choose a category 
4. Land on the confirmation page
5. Go back 
6. Close the RHP 

I prefer the first, but I'm not super passionate. Is @nkdengineer basically saying that because this is the same category selector on the confirmation page, we don't navigate back to it when you go back from there ordinarily. CC: @luacmartins @mountiny if you have any strong opinions on this.

**Comment by luacmartins** [comment]
I agree that the first option @trjExpensify laid out is better.

**Comment by mountiny** [comment]
Same as Tom and Carlos

@nkdengineer Can you please update the PR based on that expected outcome?  thanks!

**Comment by nkdengineer** [comment]
I updated with this approach. However, I noticed that the animation is different when we click on the category section of the confirmation page for the first time.

@trjExpensify cc @mountiny @Ollyws 


https://github.com/user-attachments/assets/ce12511d-0311-45b1-a4b5-c297ab01735f

**Comment by mountiny** [comment]
@nkdengineer That is not right, we should be pushing that screen in.

**Comment by nkdengineer** [comment]
@mountiny Currently, we're using `closeAndNavigate` then this bug will not happen.

https://github.com/Expensify/App/blob/37105a1b4c60ba6416b3ae3415c2265a53c55257/src/pages/iou/request/step/IOURequestStepCategory.tsx#L131


But we need to use `navigate` instead because we want this [behavior](https://github.com/Expensify/App/pull/53420#issuecomment-2521202763).

After that when we navigate to this screen, the animation is wrong because the screen is existed
https://github.com/Expensify/App/blob/37105a1b4c60ba6416b3ae3415c2265a53c55257/src/components/MoneyRequestConfirmationListFooter.tsx#L432

**Comment by mountiny** [comment]
I feel like there must be a way to achieve this @adamgrzybowski @WojtekBoman in case you have an idea on how to achieve this flow in the RHP

Yeah, my expectation was: 

1. Click "categorize it" 
2. Upgrade to a workspace to unlock categories
3. Choose a category
4. Land on the confirmation page
5. Go back 
6. Choose a category 
7. Go back 
5. Close the RHP

**Comment by WojtekBoman** [comment]
@mountiny 

I'd like to make sure. Is this what we want to achieve?



https://github.com/user-attachments/assets/e394a289-5b9d-40b9-900c-9d16f4a56053

**Comment by mountiny** [comment]
@WojtekBoman @adamgrzybowski That looks good for the flow once you already upgraded to workspace. I guess its harder when the user does not have a workspace yet and then by categorizing it, we need to create a workspace

**Comment by trjExpensify** [comment]
> I guess its harder when the user does not have a workspace yet and then by categorizing it, we need to create a workspace

They go through an upgrade interstitial. Confirming on that page (before seeing the category list) is what creates the workspace.

**Comment by WojtekBoman** [comment]
https://github.com/user-attachments/assets/a8980d33-813e-40c6-b1f3-fe5494411f59

After upgrading the workspace, categorizing flow is opened and works the same way as on the previous video that I attached

**Comment by mountiny** [comment]
@WojtekBoman nice! What changes have you made?

File: src/pages/iou/request/step/IOURequestStepCategory.tsx [hunk 1]
```diff
@@ -127,8 +127,8 @@ function IOURequestStepCategory({
 
         IOU.setMoneyRequestCategory(transactionID, updatedCategory);
 
-        if (action === CONST.IOU.ACTION.CATEGORIZE) {
-            Navigation.closeAndNavigate(ROUTES.MONEY_REQUEST_STEP_CONFIRMATION.getRoute(action, iouType, transactionID, report?.reportID ?? '-1'));
+        if (action === CONST.IOU.ACTION.CATEGORIZE && !backTo) {
+            Navigation.navigate(ROUTES.MONEY_REQUEST_STEP_CONFIRMATION.getRoute(action, iouType, transactionID, report?.reportID ?? '-1'));
```
Comment (WojtekBoman) [inline src/pages/iou/request/step/IOURequestStepCategory.tsx:131]: I implemented it the same way, but I didn't add this additional check to the if statement, I don't know what it changes

**Comment by WojtekBoman** [comment]
https://github.com/Expensify/App/blob/37105a1b4c60ba6416b3ae3415c2265a53c55257/src/components/MoneyRequestConfirmationListFooter.tsx#L432

Here, I added `CONST.NAVIGATION.ACTION_TYPE.PUSH` as the second param of `Navigation.navigate`

**Comment by WojtekBoman** [comment]
I only added the two changes I mentioned above, but I'm not very familiar with the categorization flow, so it's possible I missed something

**Comment by mountiny** [comment]
Thank you! @nkdengineer can you please check these out?

**Comment by nkdengineer** [comment]
Thanks, @WojtekBoman. @mountiny I updated the code and the test steps.

**Comment by mountiny** [comment]
Thanks! @Ollyws what is your ETA for the review?

**Review by Ollyws — APPROVED** [review]
Looks good, thanks everyone!

**Review by luacmartins — APPROVED** [review]
LGTM

**Pull request merged**